pkg/sigv4/sigv4.go

```diff
@@ -0,0 +1,254 @@
+package sigv4
+
+import (
+	"bytes"
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+)
+
+type Signer interface {
+	Sign(req *http.Request) error
+}
+
+type DefaultSigner struct {
+	config *Config
+}
+
+func NewDefaultSigner(config *Config) Signer {
+	// initialize noEscape array. This way we can avoid using init() functions
+	for i := 0; i < len(noEscape); i++ {
+		// AWS expects every character except these to be escaped
+		noEscape[i] = (i >= 'A' && i <= 'Z') ||
```

Comment:
If we populate this in a constructor, let's move this to a `Signer` property (since it's a state), and if needed, just pass downstream any method. The rationale is to avoid using a global state.